src/services/epg.service.ts

```diff
@@ -0,0 +1,114 @@
+import { array, object, string } from 'yup';
+
+import type { PlaylistItem } from '#types/playlist';
+import { getDataOrThrow } from '#src/utils/api';
+import { isValidDateString } from '#src/utils/datetime';
+import { logDev } from '#src/utils/common';
+
+export const isFulfilled = <T>(input: PromiseSettledResult<T>): input is PromiseFulfilledResult<T> => input.status === 'fulfilled';
+
+export type EpgChannel = {
+  title: string;
+  image: string;
+  programs: EpgProgram[];
+};
+
+export type EpgProgram = {
+  id: string;
+  title: string;
+  startTime: string;
+  endTime: string;
+  description?: string;
+  image?: string;
+};
+
+const epgProgramSchema = object().shape({
+  id: string().required(),
+  title: string().required(),
+  startTime: string()
+    .required()
+    .test((value) => (value ? isValidDateString(value) : false)),
+  endTime: string()
+    .required()
+    .test((value) => (value ? isValidDateString(value) : false)),
+  chapterPointCustomProperties: array().of(
+    object().shape({
+      key: string().required(),
+      value: string().required(),
+    }),
+  ),
+});
+
+class EpgService {
+  async transformProgram(data: unknown) {
+    const incomingProgram = await epgProgramSchema.validate(data);
+    const transformed: EpgProgram = {
+      id: incomingProgram.id,
+      title: incomingProgram.title,
+      startTime: incomingProgram.startTime,
+      endTime: incomingProgram.endTime,
+    };
+
+    const customProperties = incomingProgram.chapterPointCustomProperties;
+
+    // map customProperties
+    if (customProperties) {
+      transformed.image = customProperties.find((item) => item.key === 'image')?.value || '';
+      transformed.description = customProperties.find((item) => item.key === 'description')?.value || '';
+    }
+
+    return transformed;
+  }
+
+  /**
+   * Ensure the given data validates to the EpgProgram schema
+   */
+  async parseSchedule(data: unknown) {
+    if (!Array.isArray(data)) return [];
+
+    const transformResults = await Promise.allSettled(data.map((program) => this.transformProgram(program)));
+
+    return transformResults.filter(isFulfilled).map((result) => result.value);
```

Comment:
Good point

Reply:
I did this on purpose so that we only remove invalid programs from the schedule. The problem is that there are no indicators that there might be a problem. Do you have a suggestion?

Reply:
Ok, we can still change this behavior if @marcovandeveen agrees

Reply:
I'd also be fine with number 1 as is as long as we throw a logDev in for debugging

Reply:
I actually had that coded, but removed it later 😅

Reply:
Cool, let's go with that then unless Marco says otherwise